crates/optimism/rpc/src/eth/pending_block.rs

```diff
@@ -41,21 +65,192 @@ where
+            };
+
+            // add to the total blob gas used if the transaction successfully executed
+            if let Some(tx_blob_gas) = tx.blob_gas_used() {
+                sum_blob_gas_used += tx_blob_gas;
 
+                // if we've reached the max data gas per block, we can skip blob txs entirely
+                if sum_blob_gas_used == blob_params.max_blob_gas_per_block() {
+                    best_txs.skip_blobs();
+                }
+            }
+
+            // add gas used by the transaction to cumulative gas used, before creating the receipt
+            cumulative_gas_used += gas_used;
+        }
+
+        let BlockBuilderOutcome { execution_result, block, .. } =
+            builder.finish(&state_provider).map_err(Self::Error::from_eth_err)?;
+
+        let _execution_outcome = ExecutionOutcome::new(
+            db.take_bundle(),
+            vec![execution_result.receipts],
+            block.number(),
+            vec![execution_result.requests],
+        );
         let receipts = self
             .provider()
-            .receipts_by_block(block_id)?
-            .ok_or(EthApiError::ReceiptsNotFound(block_id.into()))?;
+            .receipts_by_block(block.hash().into())?
+            .ok_or(EthApiError::ReceiptsNotFound(block.hash().into()))?;
 
         Ok(Some((Arc::new(block), Arc::new(receipts))))
     }
 }
+
+pub(super) async fn track_flashblocks<N>(url: Url, state: SharedFlashblockState) -> eyre::Result<()>
+where
+    N: RpcNodeCore,
+{
+    let pubsub = WsConnect::new(url).into_service().await?;
+    let params = vec![Value::String("pending".into()), Value::Bool(true)];
+    let req: SerializedRequest =
+        Request::new(Cow::Borrowed("eth_getBlockByNumber"), Id::Number(1), params)
+            .try_into()
+            .unwrap();
+
+    loop {
+        let resp = pubsub.send(req.clone()).await?;
+        let payload = match resp.payload {
+            alloy_json_rpc::ResponsePayload::Success(p) => p,
+            alloy_json_rpc::ResponsePayload::Failure(_error_payload) => todo!(),
+        };
+        let value: serde_json::Value = serde_json::from_str(payload.get())?;
+        let new_fb: FlashBlock = serde_json::from_value(value)?;
+
+        {
+            let mut s = state.write().await;
+
+            let is_new_block = new_fb.metadata.block_number > s.metadata.block_number;
+            if is_new_block {
+                *s = new_fb;
+            } else {
+                let existing_hashes: std::collections::HashSet<_> =
+                    s.diff.transactions.iter().cloned().collect();
+
+                s.diff.transactions.extend(new_fb.diff.transactions.into_iter().filter(|tx| {
+                    let recovered: Recovered<
+    < <N as RpcNodeCore>::Primitives as reth_primitives_traits::NodePrimitives >::SignedTx
+> = recover_raw_transaction(tx).unwrap();
+
+                    !existing_hashes.contains(recovered.tx_hash().as_slice())
+                }));
```

Comment:
```suggestion
            let is_new_block = new_fb.metadata.block_number > s.metadata.block_number;
            if is_new_block {
                *state.write().await = new_fb;
            } else {
                // atomic update of transactions
                let mut s = state.write().await;
                
                let existing_raw_txns HashSet<_> = s.diff.transactions.iter().collect();
                
                let new_txns = new_fb.diff.transactions.retain(|tx| !existing_raw_txns.contains(&tx));
                
                s.diff.transactions.extend(new_txns);
```
1. always hold a write lock for the absolute minimal time possible. a good way to avoid holding on to a write lock for unnecessary time, is to avoid binding it, unless explicitly necessary to make some instructions an atomic operation.
2. it's inefficient to decode the transactions just for finding new txns. simply compare the bytes.